// vite-plugin-ssr@0.4.48/dist/cjs/node/plugin/plugins/importBuild/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.importBuild = void 0;
const plugin_1 = require("@brillout/vite-plugin-import-build/plugin");
const utils_1 = require("../../utils");
const path_1 = __importDefault(require("path"));
function importBuild() {
    let config;
    return [
        {
            name: 'vite-plugin-ssr:importBuild:config',
            enforce: 'post',
            configResolved(config_) {
                config = config_;
            }
        },
        (0, plugin_1.importBuild)({
            getImporterCode: ({ findBuildEntry }) => {
                const pageFilesEntry = findBuildEntry('pageFiles');
                return getImporterCode(config, pageFilesEntry);
            },
            libraryName: utils_1.projectInfo.projectName
        })
    ];
}
exports.importBuild = importBuild;
function getImporterCode(config, pageFilesEntry) {
    const importPathAbsolute = (0, utils_1.toPosixPath)(
    // Current file: node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/importBuild/index.js
    require.resolve(`../../../../../../dist/cjs/node/plugin/plugins/importBuild/loadBuild.js`));
    const { outDirServer } = (0, utils_1.getOutDirs)(config);
    const importPath = path_1.default.posix.relative(outDirServer, importPathAbsolute);
    const importerCode = [
        `const { setLoaders } = require('${importPath}');`,
        'setLoaders({',
        `  pageFiles: () => import('./${pageFilesEntry}'),`,
        "  clientManifest: () => require('../client/manifest.json'),",
        "  pluginManifest: () => require('../client/vite-plugin-ssr.json'),",
        '});',
        ''
    ].join('\n');
    return importerCode;
}
